storage: further optimize serializable transactions

spencerkimball opened the pull request:
This change adds two optimizations and one de-optimization.

- Because `CPut` and `Inc` now return immediately with a `WriteTooOldError`
  instead of just setting the `Transaction.WriteTooOld` flag, we no longer
  need to count them as refresh spans. Because `CPut`, at least, is a very
  commonly used KV op, this is a win. Changes to the unittests verifying
  `TxnCoordSender`-side retries show the improvements.
- Avoid retries on the slow commit path (non 1PC) in cases where there are
  no refresh spans. This requires updating the `NoRefreshSpans` flag on
  the `DistSender` if we must split a batch.
- Respect the new `Transaction.OrigTimestampWasObserved` flag to disable
  `TxnCoordSender`-side and `Replica`-side retries.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22477)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 580 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477#-L4mMTRh3ybzGBfZuzHU:-L4mMTRh3ybzGBfZuzHV:bs6lpjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97de382e35c49958ebbd534ff2815cb87cfb315/pkg/kv/dist_sender.go#L580)):*
> ```Go
> // EndTransaction. If so, the EndTransactionRequest.NoRefreshSpans
> // flag is reset to indicate whether earlier parts of the split
> // contain refersh spans.
> ```

s/refersh/refresh/

Maybe `s/contain/may create/`?

---

*[pkg/kv/dist_sender.go, line 594 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477#-L4mNksTyntfY-z7Du0G:-L4mNksTyntfY-z7Du0H:bvhrsy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97de382e35c49958ebbd534ff2815cb87cfb315/pkg/kv/dist_sender.go#L594)):*
> ```Go
> 			baPart := roachpb.BatchRequest{Requests: part}
> 			baPart.RefreshSpanIterate(nil, func(_ roachpb.Span, _ bool) {
> 				et.NoRefreshSpans = false
> ```

Is this racy? We generally try to avoid modifying the passed-in request objects. We may need to clone the EndTransactionRequest when modifying it.

---

*[pkg/kv/dist_sender.go, line 643 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477#-L4mNH3EVaAJDszScngP:-L4mNH3EVaAJDszScngQ:bhxu9pg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97de382e35c49958ebbd534ff2815cb87cfb315/pkg/kv/dist_sender.go#L643)):*
> ```Go
> 		panic("batch with MaxSpanRequestKeys needs splitting")
> 	}
> 	// To ensure that we lay down intents to prevent starvation, always
> ```

Why did this move? Now we may split the same batch twice.

---

*[pkg/roachpb/data.proto, line 365 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477#-L4mJPAKGld6hKUq92Xx:-L4mJPAKGld6hKUq92Xy:bopuwx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97de382e35c49958ebbd534ff2815cb87cfb315/pkg/roachpb/data.proto#L365)):*
> ```protobuf
>   bool orig_timestamp_was_observed = 17;
> 
>   reserved 16;
> ```

NoRefreshSpans was on EndTransactionRequest, not Transaction, so that's where we need to bump the tag number. No need to throw away tag 16 here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477#-:-L4mP4UVBttnfreV5S64:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 580 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477#-L4mMTRh3ybzGBfZuzHU:-L4mW1nlQvnhHweN8G16:b-bremmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97de382e35c49958ebbd534ff2815cb87cfb315/pkg/kv/dist_sender.go#L580)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/refersh/refresh/

Maybe `s/contain/may create/`?
</blockquote></details>

I changed it to "may result in".

---

*[pkg/kv/dist_sender.go, line 594 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477#-L4mNksTyntfY-z7Du0G:-L4mW7lkw6-Up3pPof3I:bjka1ax) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97de382e35c49958ebbd534ff2815cb87cfb315/pkg/kv/dist_sender.go#L594)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is this racy? We generally try to avoid modifying the passed-in request objects. We may need to clone the EndTransactionRequest when modifying it.
</blockquote></details>

It may be bad hygiene. Don't think it's actually racy in practice. I'll clone the `Requests` slice and add in a new `EndTransactionRequest`.

---

*[pkg/kv/dist_sender.go, line 643 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477#-L4mNH3EVaAJDszScngP:-L4mYUHPjE7utOF-eCMj:bbuvvok) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97de382e35c49958ebbd534ff2815cb87cfb315/pkg/kv/dist_sender.go#L643)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why did this move? Now we may split the same batch twice.
</blockquote></details>

Moved it back.

---

*[pkg/roachpb/data.proto, line 365 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477#-L4mJPAKGld6hKUq92Xx:-L4mYX0geF8OEahgPQO7:b-1l9j4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97de382e35c49958ebbd534ff2815cb87cfb315/pkg/roachpb/data.proto#L365)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

NoRefreshSpans was on EndTransactionRequest, not Transaction, so that's where we need to bump the tag number. No need to throw away tag 16 here. 
</blockquote></details>

Oops, changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22477)*
<!-- Sent from Reviewable.io -->

spencerkimball committed:
storage: further optimize serializable transactions

This change adds two optimizations and one de-optimization.

- Because `CPut` and `Inc` now return immediately with a `WriteTooOldError`
  instead of just setting the `Transaction.WriteTooOld` flag, we no longer
  need to count them as refresh spans. Because `CPut`, at least, is a very
  commonly used KV op, this is a win. Changes to the unittests verifying
  `TxnCoordSender`-side retries show the improvements.
- Avoid retries on the slow commit path (non 1PC) in cases where there are
  no refresh spans. This requires updating the `NoRefreshSpans` flag on
  the `DistSender` if we must split a batch.
- Respect the new `Transaction.OrigTimestampWasObserved` flag to disable
  `TxnCoordSender`-side and `Replica`-side retries.

Release note: None
